vm/vmm/vmm.go

```diff
@@ -0,0 +1,422 @@
+// Copyright 2018 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// Package vmm provides VMs based on OpenBSD vmm virtualization.
+package vmm
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func init() {
+	vmimpl.Register("vmm", ctor)
+}
+
+type Config struct {
+	Count    int    `json:"count"`    // number of VMs to use
+	CPU      int    `json:"cpu"`      // number of VM CPUs
+	Mem      int    `json:"mem"`      // amount of VM memory in MBs
+	Kernel   string `json:"kernel"`   // kernel to boot
+	Template string `json:"template"` // vm template
+}
+
+type Pool struct {
+	env   *vmimpl.Env
+	cfg   *Config
+	count int
+	mu    sync.Mutex
+}
+
+type instance struct {
+	cfg       *Config
+	image     string
+	imageID   int
+	imagePath string
+	debug     bool
+	workdir   string
+	sshkey    string
+	sshuser   string
+	sshhost   string
+	sshport   int
+	merger    *vmimpl.OutputMerger
+	vmID      int
+	wpipe     io.WriteCloser
+}
+
+func ctor(env *vmimpl.Env) (vmimpl.Pool, error) {
+	cfg := &Config{
+		Count: 1,
+		CPU:   1,
+		Mem:   512,
+	}
+
+	if !osutil.IsExist(env.Image) {
+		return nil, fmt.Errorf("image file '%v' does not exist", env.Image)
+	}
+
+	if err := config.LoadData(env.Config, cfg); err != nil {
+		return nil, fmt.Errorf("failed to parse vmm vm config: %v", err)
+	}
+	if cfg.Count < 1 || cfg.Count > 8 {
+		return nil, fmt.Errorf("invalid config param count: %v, want [1-8]", cfg.Count)
+	}
+	if env.Debug {
+		cfg.Count = 1
+	}
+	if cfg.CPU > 1 {
+		return nil, fmt.Errorf("invalid config param cpu: %v, want 1", cfg.CPU)
+	}
+	if cfg.Mem < 128 || cfg.Mem > 1048576 {
+		return nil, fmt.Errorf("invalid config param mem: %v, want [128-1048576]", cfg.Mem)
+	}
+	if cfg.Kernel == "" {
+		return nil, fmt.Errorf("missing config param kernel")
+	}
+	if !osutil.IsExist(cfg.Kernel) {
+		return nil, fmt.Errorf("kernel '%v' does not exist", cfg.Kernel)
+	}
+	if cfg.Template == "" {
+		return nil, fmt.Errorf("missing config param template")
+	}
+	pool := &Pool{
+		cfg: cfg,
+		env: env,
+	}
+
+	return pool, nil
+}
+
+func (pool *Pool) Count() int {
+	return pool.cfg.Count
+}
+
+func (pool *Pool) Create(workdir string, index int) (vmimpl.Instance, error) {
+	imageID := pool.nextID()
```

Comment:
The workdir argument passed is a per-instance temp dir. It's meant exactly for such things as a temp image file. Moreover, the original image dir can potentially be read-only. So I think we should copy the image to some fixed-name file in the workdir (e.g. `workdir/image`). This should eliminate the need in pool.count/mu. There is also a helper `osutil.CopyFile` function, no need to shell out to cp.